### PR TITLE
fix(prune) respect dir permissions during prune

### LINKED ISSUE #1872: turbo --prune --scope=[app] doesn't preserve directory file permissions
### What version of Turborepo are you using?

1.4.3

### What package manager are you using / does the bug impact?

Yarn v1

### What operating system are you using?

Mac

### Describe the Bug

I changed my `update-opensearch` directory in a yarn workspace to `chmod -R 777 update-opensearch`
```
❯ ls -al packages/workflows/src
drwxr-xr-x   9 chris  staff   288 Sep  6 11:37 .
drwxr-xr-x  13 chris  staff   416 Sep  6 00:24 ..
drwxrwxrwx   8 chris  staff   256 Sep  5 23:35 update-opensearch
drwxr-xr-x   4 chris  staff   128 Sep  1 14:56 update-trending
```

However, when I run `turbo prune --scope=workers --docker`, the directory permissions are not preserved.

```
❯ ls -al out/full/packages/workflows/src/
drwxr-xr-x  10 chris  staff   320 Sep  6 11:21 .
drwxr-xr-x  13 chris  staff   416 Sep  6 11:21 ..
drwxr-xr-x   8 chris  staff   256 Sep  6 11:21 update-opensearch
drwxr-xr-x   4 chris  staff   128 Sep  6 11:21 update-trending
```

The permissions of the text files within `update-opensearch` are preserved. I only have problems with directories.
It also doesnt work without the `--docker` flag

### Expected Behavior

I expect directory permissions to be preserved.

### To Reproduce

1. Create a turbo repo
2. Change permissions of one of the directories
3. turbo prune --scope=[app] --docker
4. check the `out` directory for permissions

### PR DESCRIPTION
**Pull request by chris-olszewski**
Fixes #1872

This PR ended up having larger scope than I originally intended. A few things that changed:
 - `RecursiveCopyFile` now attempts to copy directories with the correct permissions
 - Added `MkdirAllMode` which is similar to `MkdirAll` but it will ensure that the created directory has the permissions passed
 - Change prune to use `MkdirAllMode` instead of `EnsureDir`

## Interaction with umask
Currently we don't consider [`umask`](https://en.wikipedia.org/wiki/Umask) when creating files. When copying files that we get from users we should make sure that the permissions are kept, but when creating files for turbo itself, we should probably respect `umask`.

When copying user files and we want to make sure permissions remain the same, the standard solution is to `chmod` after creating a new dir/file. This would add an additional syscall to all of the file/directory copy methods which seems like a little much. Another option would be to [blow away `umask`](https://pkg.go.dev/syscall#Umask) at the beginning to ensure that we are creating files with the permissions we expect.
 
## Testing
Wrote some unit tests, also did a quick spot check that dir permissions are copied over
```
olszewski@chriss-mbp perm-issue % turbo_dev prune --scope=ui                                 
Generating pruned monorepo for ui in /private/tmp/perm-issue/out
 - Added ui
 - Added eslint-config-custom
 - Added tsconfig
olszewski@chriss-mbp perm-issue % ls -lh packages 
total 0
drwxr-xr-x  5 olszewski  wheel   160B Sep 22 10:43 eslint-config-custom
drwxr-xr-x  7 olszewski  wheel   224B Sep 22 10:43 tsconfig
drwxrwxrwx  7 olszewski  wheel   224B Sep 22 10:43 ui
olszewski@chriss-mbp perm-issue % ls -lh out/packages
total 0
drwxr-xr-x  5 olszewski  wheel   160B Sep 23 11:34 eslint-config-custom
drwxr-xr-x  7 olszewski  wheel   224B Sep 23 11:34 tsconfig
drwxrwxrwx  7 olszewski  wheel   224B Sep 23 11:34 ui
```

### CODE REVIEW

File: cli/internal/fs/copy_file.go [hunk 1]
```diff
@@ -24,11 +24,16 @@ func RecursiveCopy(from string, to string) error {
...
+			src := LstatCachedFile{Path: UnsafeToAbsoluteSystemPath(name), fileType: &fileType}
 			if isDir {
-				return os.MkdirAll(dest, DirPermissions)
+				mode, err := src.GetMode()
+				if err != nil {
+					return err
+				}
+				return os.MkdirAll(dest, mode)
 			}
 			// name is absolute, (originates from godirwalk)
```
Comment (gsoltis) [inline cli/internal/fs/copy_file.go:35]: I think this comment should move up to where `src` is allocated.

File: cli/internal/prune/prune.go [hunk 2]
```diff
@@ -146,10 +146,16 @@ func (p *prune) prune(opts *opts) error {
 			continue
 		}
 		workspaces = append(workspaces, ctx.PackageInfos[internalDep].Dir)
+		originalDir := p.base.RepoRoot.UntypedJoin(ctx.PackageInfos[internalDep].Dir.ToStringDuringMigration())
```
Comment (gsoltis) [inline cli/internal/prune/prune.go:149]: I think this is supposed to be `ctx.PackageInfos[internalDep].Dir.RestoreAnchor(p.base.RepoRoot)`?

@nathanhammond what do you think about also including a method on `AbsoluteSystemPath` that is essentially `Join` but accepts a single `AnchoredSystemPath` instance?

At any rate, this *shouldn't* be `Untyped`, we have the types here for both pieces of the path.

File: cli/internal/prune/prune.go [hunk 3]
```diff
@@ -146,10 +146,16 @@ func (p *prune) prune(opts *opts) error {
...
+		info, err := originalDir.Lstat()
+		if err != nil {
+			return errors.Wrapf(err, "failed to lstat %s", originalDir)
+		}
 		targetDir := fullDir.UntypedJoin(ctx.PackageInfos[internalDep].Dir.ToStringDuringMigration())
-		if err := targetDir.EnsureDir(); err != nil {
-			return errors.Wrapf(err, "failed to create folder %v for %v", targetDir, internalDep)
+		if err = targetDir.MkdirAllMode(info.Mode()); err != nil {
```
Comment (gsoltis) [inline cli/internal/prune/prune.go:155]: One advantage of this particular pattern is to scope a new `err`. By using `if err := ...`, you shadow any existing `err`, and don't leak this `err` beyond the `if` block. It's a common enough pattern that I would say you probably don't want to use it if you aren't using `:=`, or at least not without a comment explaining why. 

In this case, you aren't using the value of `err` beyond this block so I would switch to `:=`

File: cli/internal/turbopath/absolute_system_path.go [hunk 4]
```diff
@@ -124,6 +124,30 @@ func (p AbsoluteSystemPath) EnsureDir() error {
...
+	if err == nil {
+		if info.IsDir() && info.Mode() == mode {
+			// Dir exists with the correct mode
+			return nil
+		} else if info.IsDir() {
+			// Dir exists with incorrect mode
+			return os.Chmod(p.ToString(), mode)
+		} else {
+			// Path exists as file, remove it
+			if err = p.Remove(); err != nil {
```
Comment (gsoltis) [inline cli/internal/turbopath/absolute_system_path.go:139]: `:=`

File: cli/internal/turbopath/absolute_system_path.go [hunk 5]
```diff
@@ -124,6 +124,30 @@ func (p AbsoluteSystemPath) EnsureDir() error {
...
+				return err
+			}
+		}
+	}
+	if err = os.MkdirAll(p.ToString(), mode); err != nil {
```
Comment (gsoltis) [inline cli/internal/turbopath/absolute_system_path.go:144]: `:=`

**Review by gsoltis — APPROVED** [review]

**Comment by chris-olszewski** [comment]
Just notice this part of [os.Chmod](https://pkg.go.dev/os#Chmod):

> On Windows, only the 0200 bit (owner writable) of mode is used; it controls whether the file's read-only attribute is set or cleared. The other bits are currently unused.

Which explains the failing windows tests

**Review by tknickman — APPROVED** [review]

File: cli/internal/prune/prune.go (hunk 2 above)
Comment (nathanhammond) [reply cli/internal/prune/prune.go]: @gsoltis @chris-olszewski I'm game for that method; I just couldn't find a name for it that I was happy with so I didn't create it.

I agree that it isn't intuitive to start with the Anchored path and prepend the Absolute one.

**Pull request merged**